includes/utils/class-amp-image-dimension-extractor.php

```diff
@@ -150,6 +151,128 @@ private static function register_callbacks() {
 		do_action( 'amp_extract_image_dimensions_batch_callbacks_registered' );
 	}
 
+	/**
+	 * To get attachment ID from attached path.
+	 *
+	 * @param string $path Attached path.
+	 *
+	 * @return int Positive number on success, Otherwise 0.
+	 */
+	private static function get_attachment_id_from_path( $path ) {
+
+		if ( empty( $path ) ) {
+			return 0;
+		}
+
+		$path = wp_parse_url( $path, PHP_URL_PATH );
+
+		if ( empty( $path ) ) {
+			return 0;
+		}
+
+		global $wpdb;
+
+		$cache_key     = md5( $path );
+		$cache_group   = 'amp_attachment_id_from_path';
+		$attachment_id = wp_cache_get( $cache_key, $cache_group );
+
+		if ( empty( $attachment_id ) ) {
+			$attachment_id = $wpdb->get_col( // phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery
+				$wpdb->prepare(
+					"SELECT post_id FROM $wpdb->postmeta WHERE meta_key='_wp_attached_file' AND meta_value=%s ORDER BY post_id ASC;",
+					$path
+				)
+			);
+
+			$attachment_id = ( ! empty( $attachment_id ) && is_array( $attachment_id ) ) ? array_pop( $attachment_id ) : 0;
+
+			wp_cache_set( $cache_key, $attachment_id, $cache_group, 3 * HOUR_IN_SECONDS );
+		}
+
+		return (int) $attachment_id;
+	}
+
+	/**
+	 * Extract dimensions from filename if dimension exists or from file system.
+	 *
+	 * @param array $dimensions Image urls mapped to dimensions.
+	 *
+	 * @return array Dimensions mapped to image urls, or false if they could not be retrieved
+	 */
+	public static function extract_by_filename_or_filesystem( $dimensions ) {
+
+		if ( empty( $dimensions ) || ! is_array( $dimensions ) ) {
+			return [];
+		}
+
+		$upload_dir      = wp_get_upload_dir();
+		$base_upload_uri = strtolower( trim( $upload_dir['baseurl'] ) );
+
+		foreach ( $dimensions as $url => $value ) {
+
+			// Check whether some other callback attached to the filter already provided dimensions for this image.
+			if ( ! empty( $value ) && is_array( $value ) ) {
+				continue;
+			}
+
+			// Check if it's internal media or not. If it's not then bail out.
+			if ( false === strpos( strtolower( trim( $url ) ), $base_upload_uri ) ) {
+				continue;
+			}
+
+			// Get media path.
+			$attached_path = ltrim( str_replace( $base_upload_uri, '', $url ), '/' );
+
+			// Try to get attachment id from media path.
+			$attachment_id      = static::get_attachment_id_from_path( $attached_path );
+			$attachment_id      = ( ! empty( $attachment_id ) && 0 < (int) $attachment_id ) ? (int) $attachment_id : 0;
+			$possible_dimension = [];
+
+			// If attachment is exist fetch size from attachment metadata.
+			if ( ! empty( $attachment_id ) ) {
+				$possible_dimension = wp_get_attachment_metadata( $attachment_id );
+			}
```

Comment:
I'm not sure we should try to obtain the attachment post. For example, if an image URL looks like `https://example.com/wp-content/uploads/2021/04/American_bison_k5680-1-1024x668.jpg`, it will never match an attachment image because this is a generated image size. 

What's more, if the image URL does end in `-\d+x\d+\.\w+` should this bypass the need to further look up the dimensions? Could the dimensions then just be obtained from the filename itself?

Reply:
In other words, the logic you have at the  end of this function should perhaps come first, even for images that are linked to from external domains.

Reply:
I have updated the PR.

I kept the logic of getting image dimensions from the URL at last because of concerns for those images that have dimensions in the name but not actual dimensions. But this is not applicable now. Since WordPress add a suffix or those images.